apps/api/src/app/environments-v1/usecases/output-renderers/expand-email-editor-schema.usecase.ts

```diff
@@ -28,60 +33,98 @@ export class ExpandEmailEditorSchemaUsecase {
     return hydratedEmailSchema;
   }
 
-  private async processShowAndForControls(
-    variables: Record<string, unknown>,
+  private async processSpecialNodeTypes(variables: Record<string, unknown>, rootNode: TipTapNode): Promise<TipTapNode> {
+    const processedNode = structuredClone(rootNode);
+    await this.traverseAndProcessNodes(processedNode, variables);
+
+    return processedNode;
+  }
+
+  private async traverseAndProcessNodes(
     node: TipTapNode,
-    parentNode?: TipTapNode
-  ): Promise<TipTapNode> {
-    if (node.content) {
-      const processedContent: TipTapNode[] = [];
-      for (const innerNode of node.content) {
-        const processed = await this.processShowAndForControls(variables, innerNode, parentNode);
-        if (processed) {
-          processedContent.push(processed);
+    variables: Record<string, unknown>,
+    parent?: TipTapNode
+  ): Promise<void> {
+    const queue: Array<{ node: TipTapNode; parent?: TipTapNode }> = [{ node, parent }];
+
+    while (queue.length > 0) {
+      const current = queue.shift()!;
+      await this.processNode(current.node, variables, current.parent);
+
+      if (current.node.content) {
+        for (const childNode of current.node.content) {
+          queue.push({ node: childNode, parent: current.node });
         }
       }
-      node.content = processedContent;
     }
+  }
```

Comment:
Reworked the previously recursive function to something easier to read using queue.